src/cheri_ptw.sail

```diff
@@ -24,39 +24,43 @@ function ptw_error_to_str(e) =
 
 overload to_str = {ptw_error_to_str}
 
-function ext_get_ptw_error(ext_ptw : ext_ptw) -> PTW_Error =
-  match (ext_ptw) {
-    PTW_CAP_OK        => PTW_No_Permission(),
-    PTW_LOAD_CAP_ERR  => PTW_Ext_Error(AT_CAP_ERR),
-    PTW_STORE_CAP_ERR => PTW_Ext_Error(AT_CAP_ERR)
+function ext_get_ptw_error(eptwf : ext_ptw_fail) -> PTW_Error =
+  match (eptwf) {
+    EPTWF_NO_PERM  => PTW_No_Permission(),
+    EPTWF_CAP_ERR  => PTW_Ext_Error(AT_CAP_ERR)
   }
 
 /* conversion of these translation/PTW failures into architectural exceptions */
 function translationException(a : AccessType(ext_access_type), f : PTW_Error) -> ExceptionType = {
   let e : ExceptionType =
   match (a, f) {
-    /* First handle the vmem extension errors */
-    (ReadWrite(Cap), PTW_Ext_Error(AT_CAP_ERR)) => E_Extension(EXC_SAMO_CAP_PAGE_FAULT),
-    (Read(Cap), PTW_Ext_Error(AT_CAP_ERR))      => E_Extension(EXC_LOAD_CAP_PAGE_FAULT),
-    (Write(Cap), PTW_Ext_Error(AT_CAP_ERR))     => E_Extension(EXC_SAMO_CAP_PAGE_FAULT),
+    /* Tag-asserting stores can raise CHERI page faults. */
+    (Write(Cap), PTW_Ext_Error(AT_CAP_ERR))        => E_Extension(EXC_SAMO_CAP_PAGE_FAULT),
+    (ReadWrite(Cap, _), PTW_Ext_Error(AT_CAP_ERR)) => E_Extension(EXC_SAMO_CAP_PAGE_FAULT),
+    (ReadWrite(_, Cap), PTW_Ext_Error(AT_CAP_ERR)) => E_Extension(EXC_SAMO_CAP_PAGE_FAULT),
```

Comment:
Can we ever get `ReadWrite(Data, Cap)` or is it always `ReadWrite(Cap, Cap)` because we don't yet know the tag, and hence covered by the above case so should instead use the internal_error case below?

Reply:
At the moment, lacking any A extension bits for CHERI, we never generate `ReadWrite` other than `ReadWrite(Data,Data)` from upstream RISC-V, so that's a little embarrassing.  But I think we can get each of the other flavors with `amoswap.c`:

* `RW(Data, Cap)` is using an authorizing cap without `Permit_Load_Cap` (but with `Load`, `Store`, and `Store_Cap`) and storing a tagged quantity out to memory.
* `RW(Cap, Data)` is using an authorizing cap with all four relevant perms asserted and storing a tag-clear quantity to memory.
* `RW(Cap, Cap)` has all the relevant authority bits and a set tag going out to memory.

Do I have that right?

Reply:
Currently if you're missing Permit_Load_Cap you still do a Read(Cap), so it would still be a ReadWrite(Cap, Cap) or ReadWrite(Cap, Data) depending on the tag of the thing being stored. Debatable whether that should change.

Reply:
It's currently scheduled to change (inasmuch as that's a thing) in #18: it'll become `Read(if vialc then Cap else Data)` therein.

Reply:
Ok